lightning/backend/local.go

```diff
@@ -0,0 +1,1159 @@
+
+	leaderID := region.Leader.GetId()
+	clients := make([]sst.ImportSST_WriteClient, 0, len(region.Region.GetPeers()))
+	requests := make([]*sst.WriteRequest, 0, len(region.Region.GetPeers()))
+	for _, peer := range region.Region.GetPeers() {
+		cli, err := local.getImportClient(ctx, peer)
+		if err != nil {
+			return nil, err
+		}
+
+		wstream, err := cli.Write(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		// Bind uuid for this write request
+		req := &sst.WriteRequest{
+			Chunk: &sst.WriteRequest_Meta{
+				Meta: meta,
+			},
+		}
+		if err = wstream.Send(req); err != nil {
+			return nil, err
+		}
+		clients = append(clients, wstream)
+		requests = append(requests, req)
+	}
+
+	bytesBuf := newBytesBuffer()
+	defer bytesBuf.destroy()
+	pairs := make([]*sst.Pair, 0, local.batchWriteKVPairs)
+	count := 0
+	size := int64(0)
+	totalCount := 0
+	for iter.First(); iter.Valid(); iter.Next() {
+		size += int64(len(iter.Key()) + len(iter.Value()))
+		pair := &sst.Pair{
+			Key:   bytesBuf.addBytes(iter.Key()),
+			Value: bytesBuf.addBytes(iter.Value()),
+			//Key: append([]byte{}, iter.Key()...),
+			//Value:append([]byte{}, iter.Value()...),
+		}
+		if bytes.Compare(iter.Key(), startKey) < 0 {
+			log.L().Fatal("key out of order", zap.Binary("current", iter.Key()), zap.Binary("fist", startKey))
+		}
+		pairs = append(pairs, pair)
+		count++
+		totalCount++
+
+		if count >= local.batchWriteKVPairs {
+			for i := range clients {
+				requests[i].Reset()
+				requests[i].Chunk = &sst.WriteRequest_Batch{
+					Batch: &sst.WriteBatch{
+						CommitTs: engineFile.Ts,
+						Pairs:    pairs,
+					},
+				}
+				if err := clients[i].Send(requests[i]); err != nil {
+					return nil, err
+				}
+			}
+			count = 0
+			pairs = pairs[:0]
+			bytesBuf.reset()
+		}
+	}
+
+	if count > 0 {
+		for i := range clients {
+			requests[i].Reset()
+			requests[i].Chunk = &sst.WriteRequest_Batch{
+				Batch: &sst.WriteBatch{
+					CommitTs: engineFile.Ts,
+					Pairs:    pairs,
+				},
+			}
+			if err := clients[i].Send(requests[i]); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	var leaderPeerMetas []*sst.SSTMeta
+	for i, wStream := range clients {
+		if resp, closeErr := wStream.CloseAndRecv(); closeErr != nil {
+			return nil, closeErr
+		} else {
+			if leaderID == region.Region.Peers[i].GetId() {
+				leaderPeerMetas = resp.Metas
+				log.L().Debug("get metas after write kv stream to tikv", zap.Reflect("metas", leaderPeerMetas))
+			}
+		}
+	}
+
+	log.L().Debug("write to kv", zap.Reflect("region", region), zap.Uint64("leader", leaderID),
+		zap.Reflect("meta", meta), zap.Reflect("return metas", leaderPeerMetas),
+		zap.Int("kv_pairs", totalCount), zap.Int64("total_bytes", size),
+		zap.Int64("buf_size", bytesBuf.totalSize()))
+
+	return leaderPeerMetas, nil
+}
+
+func (local *local) Ingest(ctx context.Context, meta *sst.SSTMeta, region *split.RegionInfo) (*sst.IngestResponse, error) {
+
+	select {
+	case <-ctx.Done():
+		return nil, nil
+	default:
+	}
```

Comment:
i don't think we should do this. we should let `Ingest()` naturally fail with `context.Canceled`.